test/unittests/execute_floating_point_test.cpp

```diff
@@ -71,6 +71,47 @@ TEST(execute_floating_point, f64_add)
     EXPECT_THAT(execute(*instance, 0, {1.0011, 6.0066}), Result(7.0077));
 }
 
+TEST(execute_floating_point, f64_promote_f32)
+{
+    /* wat2wasm
+    (func (param f32) (result f64)
+      local.get 0
+      f64.promote_f32
+    )
+    */
+    auto wasm = from_hex("0061736d0100000001060160017d017c030201000a070105002000bb0b");
+    auto instance = instantiate(parse(wasm));
+
+    using f32_limits = std::numeric_limits<float>;
+    using f64_limits = std::numeric_limits<double>;
+
+    EXPECT_THAT(execute(*instance, 0, {0.0f}), Result(0.0));
+    EXPECT_THAT(execute(*instance, 0, {-0.0f}), Result(-0.0));
+    EXPECT_THAT(execute(*instance, 0, {1.0f}), Result(1.0));
+    EXPECT_THAT(execute(*instance, 0, {-1.0f}), Result(-1.0));
+
+    EXPECT_THAT(execute(*instance, 0, {f32_limits::max()}), Result(double{f32_limits::max()}));
+    EXPECT_THAT(execute(*instance, 0, {f32_limits::min()}), Result(double{f32_limits::min()}));
+    EXPECT_THAT(
+        execute(*instance, 0, {f32_limits::lowest()}), Result(double{f32_limits::lowest()}));
+
+    EXPECT_THAT(execute(*instance, 0, {f32_limits::infinity()}), Result(f64_limits::infinity()));
+    EXPECT_THAT(execute(*instance, 0, {-f32_limits::infinity()}), Result(-f64_limits::infinity()));
+
+    // The canonical NaN must result in canonical NaN (only the top bit set).
+    EXPECT_THAT(execute(*instance, 0, {FP32::nan()}), Result(FP64::nan()));
+
+    // An arithmetic NaN must result in any arithmetic NaN (payload >= canonical payload).
+    const auto res1 = execute(*instance, 0, {FP32::nan(FP32::canon + 1)});
+    ASSERT_TRUE(!res1.trapped && res1.has_value);
+    EXPECT_GT(FP{res1.value.f64}.nan_payload(), FP64::canon);
```

Comment:
```suggestion
    EXPECT_GE(FP{res1.value.f64}.nan_payload(), FP64::canon);
```

Reply:
Or add `is_arithmetic_nan` to `FP`

Reply:
I prefer the more restrictive check. Explanation in the comments.